[charts] Add a localization provider

alexfauquette opened the pull request:
Close #16963

This introduces the localisation for charts. It will be more useful when we introduce the toolbar since buttons and helpers will need some text.

For info, the data grid put its locales in its `apiRef` and the pickers in a dedicated context.

I went with the picker approach because we have a similar issue: defining the locale once and using it in multiple components and their subcomponent.

Having a dedicated provider allows targeting it in the MUI theme's defaultProps.

Preview: https://deploy-preview-17325--material-ui-x.netlify.app/x/react-charts/localization/

bernardobelchior commented on packages/x-charts-pro/src/BarChartPro/BarChartPro.tsx:
```diff
@@ -192,6 +192,10 @@ BarChartPro.propTypes = {
    */
   loading: PropTypes.bool,
+  /**
+   * Locale for charts components texts
```
```suggestion
   * Locale text for charts components
```

bernardobelchior commented on packages/x-charts/src/ChartsLocalizationProvider/ChartsLocalizationProvider.test.tsx:
```diff
@@ -0,0 +1,108 @@
+    const handleContextChange = spy();
+
+    render(
+      <ChartsLocalizationProvider localeText={{ noData: 'Not Priotized' }}>
+        <ChartsLocalizationProvider localeText={{ noData: 'Priotized' }}>
```
```suggestion
      <ChartsLocalizationProvider localeText={{ noData: 'Not Prioritized' }}>
        <ChartsLocalizationProvider localeText={{ noData: 'Prioritized' }}>
```

bernardobelchior commented on packages/x-charts/src/ChartsLocalizationProvider/ChartsLocalizationProvider.test.tsx:
```diff
@@ -0,0 +1,108 @@
+    );
+
+    const localeText: ChartsLocaleText = handleContextChange.lastCall.args[0].localeText;
+    expect(localeText.noData).to.equal('Priotized');
```
```suggestion
    expect(localeText.noData).to.equal('Prioritized');
```

bernardobelchior commented on packages/x-charts/src/ChartsLocalizationProvider/ChartsLocalizationProvider.test.tsx:
```diff
@@ -0,0 +1,108 @@
+    );
+
+    const localeText: ChartsLocaleText = handleContextChange.lastCall.args[0].localeText;
+    expect(localeText.noData).to.equal('Priotized');
```
```suggestion
    expect(localeText.noData).to.equal('Prioritized');
```

bernardobelchior commented on packages/x-charts/src/ChartsLocalizationProvider/ChartsLocalizationProvider.test.tsx:
```diff
@@ -0,0 +1,108 @@
+    expect(localeText.noData).to.equal('Priotized');
+  });
+
+  it("should not loose locales from higher LocalizationProvider when deepest one don't have the translation key", () => {
```
```suggestion
  it("should not lose locales from higher LocalizationProvider when deepest one don't have the translation key", () => {
```

bernardobelchior commented on packages/x-charts/src/ChartsLocalizationProvider/ChartsLocalizationProvider.tsx:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Demos:
+ *
+ * - [Date format and localization](https://mui.com/x/react-date-pickers/adapters-locale/)
+ * - [Calendar systems](https://mui.com/x/react-date-pickers/calendar-systems/)
+ * - [Translated components](https://mui.com/x/react-date-pickers/localization/)
+ * - [UTC and timezones](https://mui.com/x/react-date-pickers/timezone/)
+ *
+ * API:
+ *
+ * - [LocalizationProvider API](https://mui.com/x/api/date-pickers/localization-provider/)
```
Should we update these links?

alexfauquette replied:
Thanks, I forgot it

bernardobelchior commented on packages/x-charts/src/hooks/useChartsLocalization.ts:
```diff
@@ -0,0 +1,17 @@
+  if (localization === null) {
+    throw new Error(
+      [
+        'MUI X: Can not find the charts pickers localization context.',
```
```suggestion
        'MUI X: Can not find the charts localization context.',
```

bernardobelchior commented on packages/x-charts/src/locales/frFR.ts:
```diff
@@ -0,0 +1,10 @@
+import { getChartsLocalization } from './utils/getChartsLocalization';
+
+const frFRCharts: Partial<ChartsLocaleText> = {
+  // No group
```
What do you mean by "no group"?

alexfauquette replied:
I added one to remove this default comment

bernardobelchior commented on packages/x-charts/src/locales/utils/chartsLocaleTextApi.ts:
```diff
@@ -0,0 +1,12 @@
+   */
+  loading: string;
+  /**
+   * Title displayed in the overlay if their is no data to display.
```
```suggestion
   * Title displayed in the overlay if there is no data to display.
```

bernardobelchior commented on docs/data/charts/localization/ChartsLocalizationTableNoSnap.js:
```diff
@@ -0,0 +1,7 @@
+import * as React from 'react';
+import LocalisationTable from 'docsx/src/modules/components/LocalizationTable';
+import data from './data.json';
+
+export default function ChartsLocalizationTableNoSnap() {
+  return <LocalisationTable data={data} />;
```
```suggestion
import LocalizationTable from 'docsx/src/modules/components/LocalizationTable';
import data from './data.json';

export default function ChartsLocalizationTableNoSnap() {
  return <LocalizationTable data={data} />;
```

bernardobelchior commented on packages/x-charts/src/ChartsLocalizationProvider/ChartsLocalizationProvider.tsx:
```diff
@@ -0,0 +1,88 @@
+export interface ChartsLocalizationProviderProps {
+  children?: React.ReactNode;
+  /**
+   * Locale for charts components texts.
```
```suggestion
   * Localized text for chart components.
```

oliviertassinari commented:
How does starting to follow the docs structure for l10n in #17475 feel?

oliviertassinari commented on scripts/l10n.ts:
```diff
@@ -39,6 +39,13 @@ const packagesWithL10n = [
+  {
+    key: 'charts',
+    reportName: '📊 Charts',
+    constantsRelativePath: 'packages/x-charts/src/constants/defaultLocale.ts',
```
How does aligning all of those to the `pickers` feel? Propose in #17475

```suggestion
    constantsRelativePath: 'packages/x-charts/src/locales/enUS.ts',
```

oliviertassinari commented on packages/x-charts/src/ChartsLocalizationProvider/ChartsLocalizationProvider.tsx:
```diff
@@ -0,0 +1,88 @@
+ *
+ * - [ChartsLocalizationProvider API](https://mui.com/x/api/charts/charts-localization-provider/)
+ */
+function ChartsLocalizationProvider(inProps: ChartsLocalizationProviderProps) {
```
What's the point of having a `ChartsLocalizationProvider`? I mean, why do we need this when we have the `ThemeProvider`?

I guess this is the unstyled version. But then, why don't we have a `LocalizationProvider` component that is shared between all components?

oliviertassinari commented on docs/data/charts/localization/localization.md:
````diff
@@ -0,0 +1,115 @@
+```jsx
+import { createTheme, ThemeProvider } from '@mui/material/styles';
+import { BarChart } from '@mui/x-charts/BarChart';
+import { frFR } from '@mui/x-charts/locales';
````
Might be clearer

```suggestion
import { chartsfrFR } from '@mui/x-charts/locales';
```

alexfauquette replied:
That would break the logic  we use on all packages and their related example.

For now, all packages export locales as follow:

```jsx
import { xxYY } from '@mui/[packageName]/locales';
```

And for multiple packages, we encourage them to do
```jsx
import { xxYY as [packageName]XxYY } from '@mui/[packageName]/locales';
```

alexfauquette replied:
While doing the update, I understood your point. Exporting an object that contained only the `localeText` content.

I propose to export `frFR` and `frFRLocaleText` to be more explicit that `chartsfrFR` that does not says if the content is made for `localText` prop of the `createTheme()`

alexfauquette commented:
I tried the structure proposed in #17475

I moved the `### Built-in locales` higher in the order because I felt the need to introduce the information "We have some translation" before explaining how to use them.

I moved the table of available locales completely at the end of the page because it can grow a lot (like the [data grid one](https://mui.com/x/react-data-grid/localization/#supported-locales) ) and then cut the reading experience
